### PR TITLE
External module resolution

### LINKED ISSUE #2338: External module resolution logic
**Problem**

Current module resolution logic is roughly based on Node module loading logic however not all aspects of Node specific module loading were implemented. Also this approach does not really play well with scenarios like RequireJS\ES6 style module loading where resolution of relative files names is performed deterministically using the [base url](http://requirejs.org/docs/api.html#config-baseUrl) without needing the folder walk. Also current process does not allow user to specify extra locations for module resolution.

**Proposal**

Instead of using one hybrid way to resolve modules, have two implementations, one for out-of-browser workflows (i.e Node) and one for in-browser versions (ES6). These implementations should closely mimic its runtime counterparts to avoid runtime failures when design time module resolution succeeded and vice versa.

**Node**

Resolution logic should use the following algorithm (originally taken from [Modules all toghether](https://nodejs.org/api/modules.html#modules_all_together)):

*require(X) from module at path Y*

```javascript

If exists ambient external module named X {
  return the ambient external module 
}
else if X begins with './' or '../' or it is rooted path {
  try LOAD_AS_FILE(Y + X, loadOnlyDts=false)
  try LOAD_AS_DIRECTORY(Y + X, loadOnlyDts=false)
}
else {
  LOAD_NODE_MODULES(X, dirname(Y))
}
THROW "not found"

function LOAD_AS_FILE(X, loadOnlyDts) {
  if loadOnlyDts then load X.d.ts 
  else { 
    if  X.ts is a file, load X.ts
    else if X.tsx is a file, load X.tsx
    else If X.d.ts is a file, load X.d.ts
  }
}

function LOAD_AS_DIRECTORY(X, loadOnlyDts) {
  If X/package.json is a file {
    Parse X/package.json, and look for "typings" field.
    if parsed json has field "typings": 
    let M = X + (json "typings" field)
    LOAD_AS_FILE(M, loadOnlyDts).
  }
  LOAD_AS_FILE(X/index, loadOnlyDts)
}

function LOAD_NODE_MODULES(X, START) {
  let DIRS=NODE_MODULES_PATHS(START)
  for each DIR in DIRS {
    LOAD_AS_FILE(DIR/X, loadOnlyDts=true)
    LOAD_AS_DIRECTORY(DIR/X, loadOnlyDts=true)
  }
}

function NODE_MODULES_PATHS(START) {
  let PARTS = path split(START)
  let I = count of PARTS - 1
  let DIRS = []
  while I >= 0 {
    if PARTS[I] = "node_modules" CONTINUE
    DIR = path join(PARTS[0 .. I] + "node_modules")
    DIRS = DIRS + DIR
    let I = I - 1
  }
  return DIRS
}
```

**RequireJS/ES6 module loader**
- If module name starts with './' - then name is relative to the file that imports module or calls `require`.
- If module name is a relative path (i.e. 'a/b/c') - it is resolved using the base folder.

Base folder can be either specified explicitly via command line option or can be inferred:
- if compiler can uses 'tsconfig.json' to determine files and compilation options then location of 'tsconfig.json' is the base folder
- otherwise base folder is common subpath for all explicitly provided files

Path mappings can be used to customize module resolution process. In 'package.json' these mappings can be represented as JSON object with a following structure:

```json
  {
    "*.ts":"project/ts/*.ts",
    "annotations": "/common/core/annotations"
  }
```
Property name represents a pattern that might contain zero or one asterisk (which acts as a capture group). Property value represents a substitution that might contain zero or one asterisk - here it marks the location where captured content will be spliced. For example mapping above for a path 'assert.ts' will produce a string 'project/ts/assert.ts'. Effectively this logic is the same with the implementation of `locate` function in [System.js](https://github.com/ModuleLoader/es6-module-loader/blob/master/src/system.js#L221).

With path mappings in mind module resolution can be described as:
```typescript
for (var path in [relative_path, relative_path + '.ts', relative_path + "d.ts"]) {
    var mappedPath = apply_path_mapping(path);
    var candidatePath = isPathRooted(mappedPath) ? mappedPath : combine(baseFolder, mappedPath);
    if (fileExists(candidatePath)) {
        return candidatePath
    }
}
return undefined
```
With path mappings it becomes trivial to resolve some module names to files located on network share or some location on the disk outside the project folder.  

```json
{
    "*.ts": "project/scripts/*.ts",
    "shared/*": "q:/shared/*.ts"
}
```

Using this mapping relative path 'shared/core' will be mapped to absolute path 'q:/shared/core.ts'. 

We can apply the same resolution rules for both modules and tripleslash references though for the latter onces its is not strictly necessary since they do not implact runtime in any way.

### PR DESCRIPTION
**Pull request by Zoltu**
Fixes #2338 ... at least half of it.

*TL;DR*: See last commit, ignore others.

Note: This is branched off of https://github.com/Microsoft/TypeScript/pull/3324 and has Microsoft/TypeScript#externalModuleResolution rebased on master underneath it since that seemed to be relevant for this change.  It also has some jake commands added for ease of testing.  If this PR nears approval, I will clean it up as desired by maintainers.  In the meantime, check the last commit for the meat of the PR.

Summary
----
Adds support for module resolution via a mapping file.

This is a proof of concept for module resolution via mapping file(s).  Includes limited test coverage, no caching and poor factorization.  If the spec for this feature lands on something near this PR I don't mind fixing it up, though I don't know how to write idiomatic TSC code so the PR will still likely need some maintainer love.

CAVEAT
----
Currently only supports references within the project root directory tree.  Since dependencies managed by a package manager are likely to sit *outside* of the project root directory tree this solution falls short of achieving full desired results.  Unfortunately, I am not sure how to find a mapping file that is shipped with third party modules (outside the project root tree) since I can't walk up the directory tree indefinitely, and it is undefined how far up is a reasonable distance to walk.

This is particularly important for dependencies that have dependencies since one can reference a dependency outside of the tree with a mapping file but that dependency can't have a mapping file of its own up its directory tree.  This means that dependencies would have to either have relative file paths (no mappings) or they would need to use whatever you have defined in your project's mapping file(s).  In the first case, this hurts library distribution because dependency managers have to do the terrible NPM thing of having every dependency have its own node_modules folder.  In the second case, this just leads to broken apps when you cause a dependency to use a mis-matched version of one of its dependencies.

### CODE REVIEW

**Comment by mhegazy** [comment]
@Zoltu my apologies for the delays. @vladima is the best one to review this change as he has been looking into this for a while. but he has been pegged by a few perf and memory issues that he is investigating. i believe next week he should be able to get back to you. again sorry for the delay.

**Comment by mhegazy** [comment]
Sorry for the delay. this should be now handles by  #3147, #4154, and #4352. The change was mainly on the tooling side of the house, with VS and tsserver needing to handle resolution correctly; the previous desing had resolution responsibility split between the compiler/LS and the LS host (VS/tsserver). so that made it a bit hard to collaborate on this change. With TS 1.6, @vladima has consolidated all module resolution logic in the program, so there should be minimal host interaction into resolution, and in the future we should be able to accept similar changes without any work on the host side. Thanks again and sorry for the delay.
